### PR TITLE
Fix module dict in base finetuning

### LINKED ISSUE #8169: `ModuleDict` with `BaseFinetuning` causes recursion error
## 🐛 Bug

Using a module with `nn.ModuleDict` layers causes an infinite recursion error. This is because it passes the `Iterable` check but then iterates over keys which are strings and then subsequently pass the `Iterable` check on recursion. This is the cause of PyTorchLightning/lightning-flash#455

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #8169 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]
LGTM! Would suggest adding `ModuleDict` to the type hint for `modules`.

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
